mcp/server.go

```diff
@@ -404,13 +404,26 @@ func fileResourceHandler(dir string) ResourceHandler {
 
 // Run runs the server over the given transport, which must be persistent.
 //
-// Run blocks until the client terminates the connection.
+// Run blocks until the client terminates the connection or the provided
+// context is cancelled. If the context is cancelled, Run closes the connection.
 func (s *Server) Run(ctx context.Context, t Transport) error {
 	ss, err := s.Connect(ctx, t)
 	if err != nil {
 		return err
 	}
-	return ss.Wait()
+
+	ssClosed := make(chan error)
+	go func() {
+		ssClosed <- ss.Wait()
+	}()
+
+	select {
+	case <-ctx.Done():
+		ss.Close()
```

Comment:
This can potentially block indefinitely waiting for existing requests to finish. I think we want non-graceful shutdown in this case.

Reply:
OK, if the code was already doing it, leave it.